includes/profile.inc

```diff
@@ -449,3 +459,368 @@ function cm_tools_update_missing_default_field_values_insert_with_languages(arra
+ *   table. Conditions may be added to it to filter down the list of rows
+ *   processed. It receives these arguments:
+ *   - $query - The DB select query object,
+ *   - $field - The field info array for the current field,
+ *   - $table_type - One of CM_TOOLS_FIELD_TABLE_DATA, or
+ *                   CM_TOOLS_FIELD_TABLE_REVISION.
+ * @param callable $work_callback
+ *   The callback to apply to each row of each database table returned from the
+ *   $field_list_callback. It is called with these arguments:
+ *   - $row - Associative array of column name => values for the row. Alter by
+ *            reference,
+ *   - $field - The field info array for the current field,
+ *   - $table_type - One of CM_TOOLS_FIELD_TABLE_DATA, or
+ *                   CM_TOOLS_FIELD_TABLE_REVISION.
+ *   The $work_callback can return one of these values:
+ *   - NULL (default): Updates the row if a column was changed. Changes to
+ *                     columns used in the primary key are ignored.
+ *   - TRUE: Forces the row to be updated, even if nothing changed,
+ *   - FALSE: Skips row update, even if something changed.
+ * @param int $rows_per_pass
+ *   The number of table rows to load and process in each batch iteration.
+ * @param int $tables
+ *   Integer flags indicating which field tables to include. Defaults to both
+ *   'data' and 'revision' tables. Possible values are:
+ *   - CM_TOOLS_FIELD_TABLE_DATA -- include field data tables only,
+ *   - CM_TOOLS_FIELD_TABLE_REVISION -- include field revision tables only.
+ *
+ * @return string
+ *   A status message suitable for the return value of your update hook.
+ */
+function cm_tools_field_tables_walk(array &$sandbox, callable $field_list_filter_callback, callable $query_alter_callback, callable $work_callback, $rows_per_pass = 50, $tables = CM_TOOLS_FIELD_TABLE_DATA | CM_TOOLS_FIELD_TABLE_REVISION) {
+  // Convenience reference.
+  if (!isset($sandbox['_cm_tools_field_tables_walk'])) {
+    $sandbox['_cm_tools_field_tables_walk'] = array();
+  }
+  $s = &$sandbox['_cm_tools_field_tables_walk'];
+
+  $status = array();
+
+  // Initialize state, shared variables and gather fields.
+  if (empty($s)) {
+    $s['fields'] = call_user_func(
+      $field_list_filter_callback,
+      // We only support fields using the default SQL storage.
+      array_filter(field_info_fields(), function ($field) {
+        return $field['storage']['module'] == 'field_sql_storage' &&
+          $field['storage']['type'] == 'field_sql_storage';
+      }),
+      field_info_instances()
+    );
+    $s['fields_count'] = count($s['fields']);
+    $s['field'] = NULL;
+    $s['table_type'] = 0;
+    $tables_str = array();
+    if ($tables & CM_TOOLS_FIELD_TABLE_DATA) {
+      $tables_str[] = 'data';
+    }
+    if ($tables & CM_TOOLS_FIELD_TABLE_REVISION) {
+      $tables_str[] = 'revision';
+    }
+    $tables_str = implode($tables_str, ' and ') . (count($tables_str) > 1 ? ' tables' : ' table');
```

Comment:
The parameters for implode() should be the other way around, for compatibility in PHP 8+.

Any reason we're not using format_plural() here?

Reply:
I wrote these lines thinking that this isn't an appropriate place to use `t()`. If that's correct, then `format_plural()` would be out of the question too.

Reply:
Fair enough 👍